### PR TITLE
[fix] Fixed bugs in auto-install script #299

### LINKED ISSUE #299: [bug] Auto install fail
I tried to install on a Ubuntu 22.04 fresh install with the auto-install I got this error:

`
Traceback (most recent call last):
  File "/usr/local/bin/docker-compose", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.10/dist-packages/compose/cli/main.py", line 81, in main
    command_func()
  File "/usr/local/lib/python3.10/dist-packages/compose/cli/main.py", line 200, in perform_command
    project = project_from_options('.', options)
  File "/usr/local/lib/python3.10/dist-packages/compose/cli/command.py", line 60, in project_from_options
    return get_project(
  File "/usr/local/lib/python3.10/dist-packages/compose/cli/command.py", line 152, in get_project
    client = get_client(
  File "/usr/local/lib/python3.10/dist-packages/compose/cli/docker_client.py", line 41, in get_client
    client = docker_client(
  File "/usr/local/lib/python3.10/dist-packages/compose/cli/docker_client.py", line 124, in docker_client
    kwargs = kwargs_from_env(environment=environment, ssl_version=tls_version)
TypeError: kwargs_from_env() got an unexpected keyword argument 'ssl_version'
make: *** [Makefile:79: start] Error 1
make: Leaving directory '/opt/openwisp/docker-openwisp'
`

Also, how should I install OpenWisp? I tried the auto install, ansible, make file, and docker-compose and got none of them to work.

### PR DESCRIPTION
**Pull request by pandafy**
Updated CI to use auto-install script for spawning containers.

Closes #299

### CODE REVIEW

File: images/openwisp_websocket/Dockerfile [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # hadolint ignore=DL3007
-FROM openwisp/openwisp-base:latest
+FROM openwisp/openwisp-base:stable
```
Comment (nemesifier) [inline images/openwisp_websocket/Dockerfile:2]: now that I see this here, it's misleading here, `latest` was better.

Comment (pandafy) [reply images/openwisp_websocket/Dockerfile:2]: I think, we should use "edge"/"dev" everywhere in the code with an option to override the tag when required. If someone is trying to build the images from source, they most likely would want the versioh which is present in the code.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        run: |
+          (make develop-pythontests && make stop) || (docker compose logs && exit 1)
+        working-directory: /opt/openwisp/docker-openwisp
+        env:
+          SELENIUM_HEADLESS: 1
+
+      # the following action is equivalent to
+      # echo "$DOCKER_HUB_SECRET" | docker login --username "$DOCKER_HUB_USERNAME" --password-stdin
+      - name: Login to Docker Hub
+        if: ${{ github.event_name == 'push' && github.ref == 'refs/heads/master' }}
```
Comment (nemesifier) [inline .github/workflows/ci.yml:68]: We must execute this only if all the steps completed successfully (eg: `if: ${{ success() }}`).

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKER_HUB_USERNAME }}
+          password: ${{ secrets.DOCKER_HUB_SECRET }}
+
+      # Skip image builds and tests since they were already done
+      - name: Publish to Docker Hub
+        if: ${{ github.event_name == 'push' && github.ref == 'refs/heads/master' }}
```
Comment (nemesifier) [inline .github/workflows/ci.yml:76]: same here

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        run: |
+          make publish USER=docker.io/openwisp TAG=dev SKIP_BUILD=true SKIP_TESTS=true
+        env:
+          SELENIUM_HEADLESS: 1
+
+      - name: Login to GitLab Container Registry
+        if: ${{ github.event_name == 'push' && github.ref == 'refs/heads/master' }}
```
Comment (nemesifier) [inline .github/workflows/ci.yml:83]: same here

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        uses: docker/login-action@v1
+        with:
+          registry: registry.gitlab.com
+          username: ${{ secrets.GITLAB_DOCKER_REGISTRY_USER }}
+          password: ${{ secrets.GITLAB_DOCKER_REGISTRY_TOKEN }}
+
+      # Skip image builds and tests since they were already done
+      - name: Publish to GitLab Container Registry
+        if: ${{ github.event_name == 'push' && github.ref == 'refs/heads/master' }}
```
Comment (nemesifier) [inline .github/workflows/ci.yml:92]: same

**Review by nemesifier — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+          (GIT_BRANCH="${{ github.head_ref }}" METRIC_COLLECTION=False SKIP_PULL=true sudo -E ./deploy/auto-install.sh <<EOF
+          latest
+
+          dashboard.openwisp.org
+          api.openwisp.org
+          vpn.openwisp.org
+          administrator@openwisp.org
+
+          EOF
+          ) || cat /opt/openwisp/autoinstall.log
```
Comment (nemesifier) [inline .github/workflows/ci.yml:51]: watch out the exit code here in case of errors

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+          ref: ${{ github.event.pull_request.head.sha }}
+
+      - name: Setup testing environment
+        id: deps
+        run: |
+          sudo curl -sL -o /bin/hadolint "https://github.com/hadolint/hadolint/releases/download/v2.12.0/hadolint-Linux-x86_64"
+          sudo chmod +x /bin/hadolint
+          echo "127.0.0.1 dashboard.openwisp.org api.openwisp.org" | sudo tee -a /etc/hosts
+          # disable metric collection during builds
+          sed -i 's/METRIC_COLLECTION=True/METRIC_COLLECTION=False/' .env
```
Comment (nemesifier) [inline .github/workflows/ci.yml:26]: maybe this is redundant now

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        id: build_images
+        if: ${{ !cancelled() && steps.deps.conclusion == 'success' }}
+        run: make compose-build nfs-build
+
+      - name: Use auto-install script to start containers
+        id: start_containers
+        if: ${{ !cancelled() && steps.build_images.conclusion == 'success' }}
+        # Do not remove the blank lines from the input.
+        run: |
+          (GIT_BRANCH="${{ github.head_ref }}" METRIC_COLLECTION=False SKIP_PULL=true sudo -E ./deploy/auto-install.sh <<EOF
```
Comment (nemesifier) [inline .github/workflows/ci.yml:42]: can you pass the .env file here?

File: deploy/auto-install.sh [hunk 9]
```diff
@@ -167,6 +168,9 @@ setup_docker_openwisp() {
 		mv $ENV_USER $INSTALL_PATH/.env &>>$LOG_FILE
 		rm -rf $ENV_USER &>>$LOG_FILE
 	fi
+	if [ "$METRIC_COLLECTION" == "False" ]; then
```
Comment (nemesifier) [inline deploy/auto-install.sh:171]: if you pass the .env file, this wouldn't be necessary right?

File: build.py [hunk 10]
```diff
@@ -42,11 +37,11 @@ def get_secret_key():
...
-        randomize_key_value('DJANGO_SECRET_KEY', 'default_secret_key')
+        update_env_file('DJANGO_SECRET_KEY', 'default_secret_key')
     if 'change-database-credentials' in arguments:
-        keygen1 = get_secret_key()
+        keygen1 = get_secret_key(allow_special_chars=False)
         keygen2 = get_secret_key()
-        randomize_key_value("DB_USER", keygen1)
-        randomize_key_value("DB_PASS", keygen2)
+        update_env_file("DB_USER", keygen1)
+        update_env_file("DB_PASS", keygen2)
```
Comment (nemesifier) [inline build.py:47]:
```suggestion
        update_env_file('DB_USER', keygen1)
        update_env_file('DB_PASS', keygen2)
```

**Review by nemesifier — CHANGES_REQUESTED** [review]

**Review by nemesifier — APPROVED** [review]

**Review by nemesifier — APPROVED** [review]
LGTM! Waiting for the build to finish.

**Pull request merged**